python/ray/experimental/dag/dag_node.py

```diff
@@ -0,0 +1,217 @@
+        # Bound node options arguments, ex: func_or_class.options(num_cpus=2)
+        self._bound_options: Dict[str, Any] = {} if options is None else options
+        # Bound kwargs to resolve that's specific to subclass implementation
+        self._bound_kwargs_to_resolve: Optional[Dict[str, Any]] = kwargs_to_resolve
+        # UUID that is not changed over copies of this node.
+        self._stable_uuid = uuid.uuid4().hex
+
+    def get_args(self) -> Tuple[Any]:
+        """Return the tuple of arguments for this node."""
+
+        return self._bound_args
+
+    def get_kwargs(self) -> Dict[str, Any]:
+        """Return the dict of keyword arguments for this node."""
+
+        return self._bound_kwargs.copy()
+
+    def get_options(self) -> Optional[Dict[str, Any]]:
+        """Return the dict of options arguments for this node."""
+
+        return self._bound_options.copy()
+
+    def execute(self) -> Union[ray.ObjectRef, ray.actor.ActorHandle]:
+        """Execute this DAG using the Ray default executor."""
+        return self._apply_recursive(lambda node: node._execute_impl())
+
+    def _get_toplevel_child_nodes(self) -> Set["DAGNode"]:
+        """Return the set of nodes specified as top-level args.
+
+        For example, in `f.remote(a, [b])`, only `a` is a top-level arg.
+
+        This set of nodes are those that are typically resolved prior to
+        task execution in Ray. This does not include nodes nested within args.
+        For that, use ``_get_all_child_nodes()``.
+        """
+
+        children = set()
+        for a in self.get_args():
+            if isinstance(a, DAGNode):
+                children.add(a)
+        for a in self.get_kwargs().values():
+            if isinstance(a, DAGNode):
+                children.add(a)
+        return children
+
+    def _get_all_child_nodes(self) -> Set["DAGNode"]:
+        """Return the set of nodes referenced by the args of this node.
+
+        For example, in `f.remote(a, [b])`, this includes both `a` and `b`.
+        """
+
+        f = _PyObjScanner()
+        children = set()
+        for n in f.find_nodes(
+            [self._bound_args, self._bound_kwargs, self._bound_options]
+        ):
+            children.add(n)
+        return children
+
+    def _replace_all_child_nodes(self, fn: "Callable[[DAGNode], T]") -> "DAGNode":
+        """Replace all immediate child nodes using a given function.
+
+        This is a shallow replacement only. To recursively transform nodes in
+        the DAG, use ``_apply_recursive()``.
+
+        Args:
+            fn: Callable that will be applied once to each child of this node.
+
+        Returns:
+            New DAGNode after replacing all child nodes.
+        """
+
+        replace_table = {}
+
+        # Find all first-level nested DAGNode children in args.
+        f = _PyObjScanner()
+        children = f.find_nodes(
+            [self._bound_args, self._bound_kwargs, self._bound_options]
+        )
```

Comment:
this should share code with `self._get_all_child_nodes`

Reply:
i just gave it a try but find it slightly less clean given the current setup. So `_get_all_child_nodes` is only used in testing with the assumption that it always creates a new scanner object, but in replace_all_nodes we want to ensure we always use the same scanner for both find & replace.

Reply:
At a minimum you can probably do:
```python
def _find_bound_nodes(self, scanner: _PyObjScanner):
    return s.find_nodes(
            [self._bound_args, self._bound_kwargs, self._bound_options]
        )
```

Reply:
yep i've tried that but makes some tests a bit more ugly and breaks a few assumptions it had regarding shared scanner :/ Let me put it as follow up item to clean up these functions with tests all together.